### PR TITLE
Lost fragments table when update object in Cache

### PR DESCRIPTION
**Pull request by taaotao007**
when set 404 response cached, first time got 404 response, then next request got 200 response,but fragment offset did not changed, this cause range request  data error.

### CODE REVIEW

**Comment by randall** [comment]
[approve ci]

File: iocore/cache/CacheWrite.cc [hunk 1]
```diff
@@ -106,7 +106,13 @@ CacheVC::updateVector(int /* event ATS_UNUSED */, Event * /* e ATS_UNUSED */)
          data should remain valid.
       */
       if (alternate_index >= 0) {
-        alternate.copy_frag_offsets_from(write_vector->get(alternate_index));
+        /*if length changed, fragment offset also need to change, if not cause range request error*/
+        if (total_len != write_vector->get(alternate_index)->object_size_get()) {
```
Comment (scw00) [inline iocore/cache/CacheWrite.cc]: Sometime we were try to update the header only fragment that means we won't write any fragment in it. So the `total_len` is always 0 in this situation.  And we will lost the fragments info forever if we don't copy the fragments table from old info.

Comment (taaotao007) [reply iocore/cache/CacheWrite.cc]: There is situation that first response code is 404 and next response code is 200,so the total_len has changed to non zero.In this situation,this cause data read error.

Comment (scw00) [reply iocore/cache/CacheWrite.cc]: But it affect the normal requests.

Comment (taaotao007) [reply iocore/cache/CacheWrite.cc]: You can test this situation.

**Comment by zwoop** [comment]
Besides fixing the build problems, can you also please make a more descriptive (but < 50 character) Subject line for this PR? Summarizing the more descriptive text in the commit.

**Comment by zwoop** [comment]
[approve ci].

**Comment by shinrich** [comment]
[approve ci]

**Comment by scw00** [comment]
[approve ci]

**Comment by scw00** [comment]
I won't approve this pr because it break the header update only case. As I comment above,  some header update object will lost the fragments table forever.

For your case. My question is:
 - How did you cache HTTP with negative code ?
 - How did you update the negative HTTP with 200 HTTP ?
 - How to reproduce that ?

**Comment by taaotao007** [comment]
> I won't approve this pr because it break the header update only case. As I comment above, some header update object will lost the fragments table forever.
> 
> For your case. My question is:
> 
> * How did you cache HTTP with negative code ?
> * How did you update the negative HTTP with 200 HTTP ?
> * How to reproduce that ?

1 jush set proxy.config.http.negative_caching_enabled 1
2 when client reqests, may be the resource just producing,at this time responsed with a 404 status,after the data is produced, the client requests get a 200 status.

**Comment by scw00** [comment]
Thanks.

So you mean the 404 is stale and ats goes back to the OS. And then OS return 200 to ats. That make sense, will test it in this weekend.

**Comment by taaotao007** [comment]
> Thanks.
> 
> So you mean the 404 is stale and ats goes back to the OS. And then OS return 200 to ats. That make sense, will test it in this weekend.

yes

**Comment by scw00** [comment]
Yes I see.

In fact, it is not only negative cache issue but also the whole cache update cases I think . When some want to update an object with response data (not header only cases), it always copy fragments table from old info. 

https://github.com/apache/trafficserver/blob/88b5a13ecb18c741c04a39b611566427c94546a4/iocore/cache/CacheWrite.cc#L108-L112

How about changing the L108 to `if (alternate_index >= 0 && total_len == 0 && alternate.get_frag_offset_count() == 0)`

like that
```
+      // As comment above, we might update header only without writting any data in cache.
+      // In that case, the new object need to copy fragments table from old info. And write it done.
+      // But assume that we update a small data to a large data. we should not copy anything, because
+      // the small data's fragments table is not suit to large one's. No doubt that allow_empty_doc is also
+      // match this condition.
+      if (alternate_index >= 0 && total_len == 0 && alternate.get_frag_offset_count() == 0) {
         alternate.copy_frag_offsets_from(write_vector->get(alternate_index));
       }

```

**Comment by scw00** [comment]
@taaotao007 Do you want to update your PR ?

**Comment by scw00** [comment]
Thanks to your report. open an issue #5694 to trace it . Please feel free to reopen it.

**Pull request closed without merging**